File: memoization/python/04_how_sum.py
"""How sum.

This script implements a solution to the how sum problem using memoization.

Given an integer target_sum and a list of numbers as arguments, return a list containing
any combination of elements that add up to exactly the target_sum. If there is no
combination that adds up to the target_sum, then return None.
"""
from typing import List, Union


def how_sum(
    target_sum: int, numbers: List[int], memo: dict = None
) -> Union[List[int], None]:
    """Find a combination of elements from numbers that add up to exactly the target_sum.

    Args:
        target_sum (int): Value to be generated.
        numbers (List[int]): List of numbers to be used to generate the target sum.
        memo (dict): A hashmap to cache combinations already calculated.

    Returns:
        Union[List[int], None]: A list containing any combination of elements that add
        up to exactly the target_sum, or None in case there is no combination that adds
        up.
    """
    if memo is None:
        memo = {}

    if target_sum in memo:
        return memo[target_sum]

    if target_sum == 0:
        return []

    if target_sum < 0:
        return None

    for number in numbers:
        remainder = target_sum - number
        result = how_sum(remainder, numbers, memo)
        if result is not None:
            result.append(number)
            memo[target_sum] = result
            return result

    memo[target_sum] = None
    return None


if __name__ == "__main__":
    print(how_sum(7, [2, 3]))  # [3,2,2]
    print(how_sum(7, [5, 3, 4, 7]))  # [4,3]
    print(how_sum(7, [2, 4]))  # None
    print(how_sum(8, [2, 3, 5]))  # [2,2,2,2]
    print(how_sum(300, [7, 14]))  # None

File: memoization/golang/04_how_sum.go
// How sum.

// Implementation of a solution to the how sum problem using memoization.

// Given an integer targetSum and a list of numbers as arguments, return a list containing
// any combination of elements that add up to exactly the targetSum. If there is no
// combination that adds up to the targetSum, then return None.
package main

import "fmt"

// Find a combination of elements from numbers that add up to exactly the targetSum.
func HowSum(targetSum int, numbers []int, memo *map[int][]int) []int {
	if combination, ok := (*memo)[targetSum]; ok {
		return combination
	}

	if targetSum == 0 {
		return []int{}
	}

	if targetSum < 0 {
		return nil
	}

	for _, number := range numbers {
		remainder := targetSum - number
		combination := HowSum(remainder, numbers, memo)
		if combination != nil {
			combination = append(combination, number)
			(*memo)[targetSum] = combination
			return combination
		}
	}

	(*memo)[targetSum] = nil
	return nil
}

func main() {
	// Must create a map for each run of HowSum, because golang has no default value for
	// function params or optional params in function. If we pass the same map as param
	// to each run of HowSum, the values stored in the first run will affect the
	// subsequent ones because not all HowSum calls have the same targetSum. Another
	// solution would be to create just one map and after each run delete all results
	// stored inside it.
	memo1 := make(map[int][]int)
	memo2 := make(map[int][]int)
	memo3 := make(map[int][]int)
	memo4 := make(map[int][]int)
	memo5 := make(map[int][]int)

	fmt.Println(HowSum(7, []int{2, 3}, &memo1))       // [3,2,2]
	fmt.Println(HowSum(7, []int{5, 3, 4, 7}, &memo2)) // [4,3]
	fmt.Println(HowSum(7, []int{2, 4}, &memo3))       // []
	fmt.Println(HowSum(8, []int{2, 3, 5}, &memo4))    // [2,2,2,2]
	fmt.Println(HowSum(300, []int{7, 14}, &memo5))    // []
}
